coding/python_testing.rst

```diff
@@ -4,6 +4,156 @@ Python Unit Testing
+It is also possible to indicate that a particular test is expected to fail, to the extent that it is reported it a test unexpectedly passes.
+Expected failures can be used to write test code that triggers a reported bug before the fix to the bug has been implemented and without causing the continuous integration system to die.
+One of the advantages of using a modern test runner such as `pytest`_ is that it is very easy to generate machine-readable pass/fail/skip/xfail statistics to see how the system is evolving over time.
+
+Using a shared base class
+=========================
+
+For some tests it is helpful to provide a base class and then share it amongst multiple test classes that are configured with different attributes.
+If this is required be careful to not have helper functions prefixed with ``test`` and do not have the base class named with a ``Test`` prefix and ensure it does not inherit from :class:`~unittest.TestCase`, if you do `pytest`_ will attempt to find tests inside it.
+Historically LSST code has dealt with this by creating a test suite that only includes the classes to be tested, omitting the base class.
+This can not work in a `pytest`_ environment, so instead multiple inheritance should be used such that the test classes inherit from :class:`~unittest.TestCase` and the support base class.
+
+Consider the following test code:
+
+.. literalinclude:: unit_test_snippets/unittest_baseclass.py
+   :language: python
+
+which runs a single test and does not attempt to run any tests in :class:`BaseClass`.
+
+.. code-block:: shell
+
+   $ py.test -v coding/unit_test_snippets/unittest_baseclass.py
+   ======================================= test session starts ========================================
+   platform darwin -- Python 3.4.3, pytest-2.9.1, py-1.4.30, pluggy-0.3.1 -- /usr/local/bin/python3.4
+   cachedir: coding/unit_test_snippets/.cache
+   rootdir: coding/unit_test_snippets, inifile:
+   collected 1 items
+
+   coding/unit_test_snippets/unittest_baseclass.py::ThisIsTest1::testParam PASSED
+
+   ===================================== 1 passed in 0.02 seconds =====================================
+
+
+LSST Utility Test Support Classes
+=================================
+
+:mod:`lsst.utils.tests` provides several utilities functions and classes for writing Python tests that developers should make use of.
+
+Special Asserts
+---------------
+
+Inheriting from :class:`lsst.utils.tests.TestCase` rather than :class:`unittest.TestCase` enables two new asserts that are useful for doing element-wise comparison of two :mod:`numpy` arrays.
+:meth:`~lsst.utils.tests.TestCase.assertClose` asserts that the corresponding elements in two arrays are within the specified tolerance.
+:meth:`~lsst.utils.tests.TestCase.assertNotClose` asserts that the two arrays are different.
```

Comment:
Yes, just like functors that register themselves in a registry. Importing afw will make these particular asserts appear. Since they test afw classes they live in afw. It's not an ideal system, but it makes sense. If you are trying to test afw objects changes are you have imported afw (directly or indirectly). I'm not sure if any others are added; certainly afw is likely to do the most because it has lots of smallish objects that can usefully be compared, but are not trivial to compare.